Sources/SwiftProtobuf/SwiftProtobufError.swift

```diff
@@ -0,0 +1,1098 @@
+        public var line: Int
+
+        public init(function: String, file: String, line: Int) {
+            self.function = function
+            self.file = file
+            self.line = line
+        }
+
+        @usableFromInline
+        internal static func here(
+            function: String = #function,
+            file: String = #fileID,
+            line: Int = #line
+        ) -> Self {
+            return SourceLocation(function: function, file: file, line: line)
+        }
+    }
+}
+
+extension SwiftProtobufError: CustomStringConvertible {
+    public var description: String {
+        "\(self.code) (at \(self.location)): \(self.message)"
+    }
+}
+
+extension SwiftProtobufError: CustomDebugStringConvertible {
+    public var debugDescription: String {
+        "\(String(reflecting: self.code)) (at \(String(reflecting: self.location))): \(String(reflecting: self.message))"
+    }
+}
+
+// - MARK: Common errors
+
+extension SwiftProtobufError {
+    /// Errors arising from encoding protobufs into binary data.
+    public enum BinaryEncoding {
+        /// The definition of the message or one of its nested messages has required
+        /// fields but the message being encoded did not include values for them. You
+        /// must pass `partial: true` during encoding if you wish to explicitly ignore
+        /// missing required fields.
+        public static func missingRequiredFields(
+            function: String = #function,
+            file: String = #fileID,
+            line: Int = #line
+        ) -> SwiftProtobufError {
+            SwiftProtobufError(
+                code: .binaryEncodingError,
+                message: """
+                The definition of the message or one of its nested messages has required fields, \
+                but the message being encoded did not include values for them. \
+                Decode with `partial: true` if you wish to explicitly ignore missing required fields.
+            """,
+                location: SourceLocation(function: function, file: file, line: line)
+            )
+        }
+        
+        /// Messages are limited to a maximum of 2GB in encoded size.
+        public static func tooLarge(
+            function: String = #function,
+            file: String = #fileID,
+            line: Int = #line
+        ) -> SwiftProtobufError {
+            SwiftProtobufError(
+                code: .binaryEncodingError,
+                message: "Messages are limited to a maximum of 2GB in encoded size.",
+                location: SourceLocation(function: function, file: file, line: line)
+            )
+        }
+        
+        /// `Any` fields that were decoded from JSON cannot be re-encoded to binary
+        /// unless the object they hold is a well-known type or a type registered via
+        /// `Google_Protobuf_Any.register()`.
+        public static func anyTypeURLNotRegistered(
+            typeURL: String,
```

Comment:
I was going back through some of our internal code, and realized this new model for errors has a limitation when it comes to the general idea of details about an error.  So in the enum form, the `typeURL` was an associated value, so a developer can write code to extract what the type was and then act on it (log it, attempt to do something else about it, etc.)

In this new form, there is no way a developer can write code to actually get at this detail to do anything with it. If they describe the error all they get is the `Code`. So if someone wanted some degree of logging for how an application does in the field, how would they get these sorta information?  Snd I don't think we want the `message` expose for the previously mentioned reasons, as then folks will parse it making the string contents part of the api contract.

Are we headed down a path where we need something like `NSError`'s `userInfo` to do this completely generically to avoid future api issues? 😞

Reply:
Quick follow up - I'm thinking cases like where say you have a server accepting connection, being able to get specific details about all the different decoding failures that are happening may help you better understand what if someone is trying to hack into your server vs. say just doing a DOS attack, etc.  So I'm starting to wonder if these "domains" are just too limited of info to provide something detailed enough to build monitoring on top of.

Reply:
If users wanted to log the `typeURL` or any other information associated with the error, they can just get the error description and the `message` will be a part of it (which now makes me think it may make having made `message` internal moot, but let's stay focused for a minute :)).
So let's assume that the "something else" users want to do with these associated values is try to recover from the error in some way, and that they need more information to do so. It could be argued that maybe throwing errors to drive the program's logic isn't necessarily the best pattern. Using this `anyTypeURLNotRegistered` as an example: if it was a recoverable issue, you might want to add API to allow users to set a closure/have a delegate to handle "unregistered URLs" for instance.

However, I do understand the concern and I think it's still valid (seeing how changing the APIs in a more significant way would either be breaking or definitely out of scope). 

So thinking of solutions that would work for this... Having struct-based errors **does** allow for associated values, but it also requires having a more specific wrapped private enum instead of this more generic "domain" approach. We could add more specific error codes as a compromise for errors that definitely required associated values. 
For instance, we could have something like:

```swift
struct SwiftProtobufError: Error {
  private enum _Error {
    case malformedAnyField
    case anyTypeURLNotRegistered(typeURL: String)
    // There would obviously be more cases here, just writing these two for simplicity
  }

  private let _wrapped: _Error
  private init(_ wrapped: _Error) {
    self._wrapped = wrapped
  }

  public static let malformedAnyField = SwiftProtobufError(.malformedAnyField)

  public static func anyTypeURLNotRegistered(typeURL: String) -> SwiftProtobufError {
    SwiftProtobufError(.anyTypeURLNotRegistered(typeURL: typeURL))
  }

  public var unregisteredTypeURL: String? {
    switch self._wrapped {
      case .malformedAnyField:
        return nil
      case .anyTypeURLNotRegistered(let typeURL):
        return typeURL
    }
  }
}
```

Callers could then match the error and extract the associated value like this:
```swift
do {
  try somethingThatThrows()
} catch let error as SwiftProtobufError {
  if let unregisteredTypeURL = error.unregisteredTypeURL {
    // This error was thrown because there was an issue with an unregistered type URL:
    // do something with it.
  }
}
```

I know this isn't as ideal, but I think it's a good compromise. We've reached this current iteration of errors as structs after quite a lot of experimentation on our other libraries, and we think it's the best we can currently do.

What do you think? Would you want to have a getter for all associated values that were part of some error in the old model?

Reply:
So it seems we'd only need two cases with associated values (and thus two getters): `anyTypeURLNotRegistered` and `unknownField`. Given the context, I think it's fine to compromise and go down this route so I'll just implement it, it's quick anyways. Others can say what they think once it's done and we can change it once again if need be.